internal/app/orchestrator/downstream.go

```diff
@@ -14,82 +14,73 @@ package orchestrator
 import (
 	"sync"
 
-	"github.com/envoyproxy/xds-relay/internal/app/mapper"
 	"github.com/envoyproxy/xds-relay/internal/app/transport"
 )
 
 // downstreamResponseMap is a map of downstream xDS client requests to response
 // channels.
 type downstreamResponseMap struct {
 	mu      sync.RWMutex
-	watches map[transport.Request]transport.Watch
+	watches map[string][]transport.Watch
 }
 
 func newDownstreamResponseMap() downstreamResponseMap {
 	return downstreamResponseMap{
-		watches: make(map[transport.Request]transport.Watch),
+		watches: make(map[string][]transport.Watch),
 	}
 }
 
 // createWatch initializes a new channel for a request if it doesn't already
 // exist.
-func (d *downstreamResponseMap) createWatch(req transport.Request) transport.Watch {
+func (d *downstreamResponseMap) addWatch(aggregatedKey string, w transport.Watch) {
 	d.mu.Lock()
 	defer d.mu.Unlock()
-	if _, ok := d.watches[req]; !ok {
-		d.watches[req] = req.CreateWatch()
+	if _, ok := d.watches[aggregatedKey]; !ok {
+		d.watches[aggregatedKey] = make([]transport.Watch, 0, 10)
 	}
-	return d.watches[req]
+	d.watches[aggregatedKey] = append(d.watches[aggregatedKey], w)
 }
 
 // get retrieves the channel where responses are set for the specified request.
-func (d *downstreamResponseMap) get(req transport.Request) (transport.Watch, bool) {
+func (d *downstreamResponseMap) getSnapshot(aggregatedKey string) ([]transport.Watch, bool) {
 	d.mu.RLock()
 	defer d.mu.RUnlock()
-	channel, ok := d.watches[req]
-	return channel, ok
+	watches, ok := d.watches[aggregatedKey]
+	if ok {
```

Comment:
Should this be `if !ok {`? If yes, may want to add a test that captures this use case.

Reply:
this is expected..if there was an entry, the snapshot is replaced. If its not found, then no action needed. Added tests for it